units/src/amount.rs

```diff
@@ -186,18 +223,86 @@ impl std::error::Error for ParseAmountError {
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
         use ParseAmountError::*;
 
-        match *self {
-            Negative | TooBig | TooPrecise | InvalidFormat | InputTooLarge
+        match self {
+            TooPrecise | InvalidFormat | InputTooLarge
             | InvalidCharacter(_) => None,
-            InvalidDenomination(ref e) => Some(e),
+            OutOfRange(error) => Some(error),
```

Comment:
@apoelstra do you want to please weigh in with the argument you made to me a year or so ago about why you think using `*self` and `ref` is better. Save me butchering your argument.

Reply:
I personally consider `ref` to be mostly old style except for rare borrow checker limitations. Just using `&self` and not sprinkling `ref` everywhere is less annoying. Also this is used in most codebases I've seen so I expect newbies to be more familiar with it.

IIUC the argument for `ref` is that it's more accurate description of what the types are. It's not wrong but I think the fact that the matching on references is obvious (which was also the reason it got implemented; yes, I was around before it was available) makes it non-issue.

Reply:
> @apoelstra do you want to please weigh in with the argument you made to me a year or so ago about why you think using `*self` and `ref` is better. Save me butchering your argument.

I prefer my code's semantics to be explicit. It saves nearly zero time when programming for the compiler to implicitly screw around with levels of dereferencing and literally zero time when reviewing (it may actually increase review time because it eliminates visual hints as to the types of various objects).

But since the compiler refuses to give you any help in noticing when it's inserting implicit `ref`s and `&`s and `*`s, I've given up on trying to enforce this by hand. There's also no clippy lint for it.

Reply:
With that said, and since we (I) have, over the last year, nearly completely used the `*self` and `ref` style codebase wide can we agree to keep it @Kixunil ?

Reply:
I'd rather not bother with it in general but since this is top commit and it makes diff smaller I changed it.